src/k8s_autoscale/slo.py

```diff
@@ -1,23 +1,31 @@
 import math
 
 
-def get_new_worker_count(pending, running, args):
+def get_target_replica_count(pending, running, args):
     # TODO: verify all the args
     assert args["slo_seconds"] > args["avg_task_duration"]
     # In case we don't want to cover all the pending tasks
     pending = int(math.ceil(pending * args["capacity_ratio"]))
     # Scale down only when we have no pending tasks
-    if pending == 0:
-        return -running
+    if pending < 1:
+        return 0
+    # Assume that all running workers have a task
+    outstanding = pending + running
     # How many tasks a replica can process within our tolerance period
-    new_tasks_per_replica = math.floor(args["slo_seconds"] / args["avg_task_duration"])
+    tasks_per_replica = math.floor(args["slo_seconds"] / args["avg_task_duration"])
     # how many tasks can be covered by the running replicas, assuming they are
     # busy and can only take new tasks after they are done with the current one
-    running_tasks_per_replica = math.floor(args["slo_seconds"] / args["avg_task_duration"]) - 1
-    running_can_cover = running * running_tasks_per_replica
-    still_pending = pending - running_can_cover
-    if still_pending > 0:
-        new_replicas_needed = math.ceil(still_pending / new_tasks_per_replica)
-        return min([new_replicas_needed, args["max_replicas"]])
-    else:
-        return 0
+    # target_replicas = running + new_replicas_needed
+    #                 = running + (still_pending / tasks_per_replica)
+    #                 = running + ((pending - running_can_cover) / tasks_per_replica
+    #                 = running + ((pending - (running*running_tasks_per_replica)) / tasks_per_replica)
+    #                 = running + ((pending - (running*(tasks_per_replica-1))) / tasks_per_replica)
+    #                 = running + ((pending + running - running*tasks_per_replica) / tasks_per_replica)
+    #                 = running + ((pending + running) / tasks_per_replica - running)
+    #                 = (pending + running) / tasks_per_replica
+    needed_replicas = math.ceil(outstanding / tasks_per_replica)
+    # Do not scale down in case we have pending, because some workers will
+    # receive SIGUSR1 and won't take any tasks after that.
+    needed_replicas = max(needed_replicas, running)
```

Comment:
Gah. I lost a long comment I had on this line, suggesting we get rid of clamping here, as I `needed_replicas` is already an overestimate when `needed_replicas < running` (assuming the workers complete the tasks that they are running before shutting down), so clamping it makes our estimate even worse.

Reply:
It makes a lot of sense to scale down when you have the capacity, but I'm worried about the situation when the polling interval is shorter than the task duration.

For example, with avg 60 and SLO 300, we have the following situation. Assuming that some of the running tasks are very close to their end and get shut down, some are not, so we don't take any new tasks, and the polling happens very often.

poll 1: pending: 10, running: 20 -> suggest to scale down to 6
poll 2: pending: 10, running:  6 -> suggest to scale down to 4
poll 3: pending: 10, running:  4 -> suggest to scale down to 3
poll 4: pending: 10, running:  3 -> noop

Also, with this simulation it makes sense to go to 3 in the first run, hmmm. Maybe I should rethink the approach.

Reply:
So, the value I got for value I arrived at for hitting our SLO with the least slack in the case we are discussing was `pending / (tasks_per_replica - 1)` which gives the value 3 you have. That said, I don't think scaling down slightly slower that is optimal is a problem, as we end up finishing work faster.

That last argument could also be used to argue for keeping the current clamping you have. If we did that, we should adjust the comment here to say that we don't scale down in order to rush to finish, rather than what is there currently.

However, in practice, given that `tasks_per_replica` (at least in prod) is always 2, any of values will lead to finishing in about the same time (since `needed_replicas >= pending` for any of the calculations). That suggest we should use the simplest logic, which I think is dropping any special handling of the the `needed_replicas < running` (either clamping, or using an alternative formula).